department-of-veterans-affairs/caseflow
# TYLERB/APPEALS-12467: Modify decision review queue search functionality to account for pagination

Resolves [APPEALS-12467](https://vajira.max.gov/browse/APPEALS-12467)
Also resolves [APPEALS-12466](https://vajira.max.gov/browse/APPEALS-12466)

### Description
Updated the decision review queue searching functionality to work with backend pagination.

### Acceptance Criteria
- [ ] Existing searching functionality is successfully ported from frontend to backend

### Testing Plan
#### Generate Test Data
1. Generating testing data can be accomplished in one of two ways
1a. Download my version of a gzipped database and follow the instructions below to get that data. The testing plan results will be based around my data
Download my gzip data from https://github.com/TylerBroyles/misc_files/blob/main/caseflow-db-backup.gz and place it in the top level of your caseflow directory
```
#Add this command to your make file. You might have to adjust the file location that it uses.
db-restore: ## Restores main Caseflow db from caseflow-db-backup.gz
	bundle exec rake db:drop
	gzip -dck caseflow-db-backup.gz | docker exec -i appeals-db psql -U postgres
```
1b. Generate your own data. This data will be mostly random so all results will be random as well and you will have to determine the correct results on your own.
```ruby
# Save the VHA business line to a variable for later use
vha = BusinessLine.find_by(url: "vha")

# Check count of open VHA business line tasks before proceeding:
vha.tasks.open.size

#
# Section 1
# Everything in the below section should always appear in the in_progress_tasks result set.
#
active_hlr_tasks = 100.times.map { FactoryBot.create(:higher_level_review_vha_task) }
active_sc_tasks = 100.times.map { FactoryBot.create(:supplemental_claim_vha_task) }
active_vrr_tasks = 100.times.map { FactoryBot.create(:veteran_record_request_task, assigned_to: vha) }
active_vrr_tasks.each { |t| FactoryBot.create(:request_issue, decision_review: t.appeal) }

#
# Section 2
# Everything in the below section should never appear in the in_progress_Tasks result set.
#
completed_hlr_tasks = 100.times.map { FactoryBot.create(:higher_level_review_vha_task) }.each(&:completed!)
completed_sc_tasks = 100.times.map { FactoryBot.create(:supplemental_claim_vha_task) }.each(&:completed!)
completed_vrr_tasks = 100.times.map { FactoryBot.create(:veteran_record_request_task, assigned_to: vha) }.each(&:completed!)
open_vrr_tasks_without_ris = 100.times.map { FactoryBot.create(:veteran_record_request_task, assigned_to: vha) }

#
# Section 3
# These BoardGrantEffectuationTasks (BGTs) will appear if the board_grant_effectuation_task
# FeatureToggle is enabled
#
bgt_tasks = 100.times.map { FactoryBot.create(:board_grant_effectuation_task, assigned_to: vha) }
bgt_tasks.each do |t|
  issue = FactoryBot.create(:request_issue, decision_review: t.appeal)
  issue.update!(closed_at: Time.zone.now, closed_status: "decided")
end
```
#### Start of testing steps
1. Login as a user capable of viewing the decision review queue for vha. I typically use the intake user shaw with css_id: BVAISHAW
2. Go to the vha decision review page at the url /decision_review/vha
3. If you have my data setup correctly using the gzip the initial page should look something like the screenshot below
![image](https://user-images.githubusercontent.com/109369527/210838895-a6faf443-8346-4c12-aa35-e2c0725678ee.png)

4. Search for the text Kara and there should only be one record that matches also search for kara to see the same results to verify that the search is not case sensitive 
![image](https://user-images.githubusercontent.com/109369527/210839251-dea8f2f8-06c4-458d-823e-f9805d656219.png)
5. Search for the text trem and there should be approximately 14 records including a board grant
![image](https://user-images.githubusercontent.com/109369527/210839858-bc5d7726-1689-450e-ad9d-7ec4bca2eb4a.png)

6.  Go into your rails console and disable the board effectuation grant task feature toggle
```ruby
FeatureToggle.disable!(:board_grant_effectuation_task)
```
7. Reload the page and search for trem again and there should only be 13 records now excluding the board grant
![image](https://user-images.githubusercontent.com/109369527/210840139-62bcee75-b80d-487d-963b-77d0b8832607.png)

8. Reenabled the feature toggle if you want
```ruby
FeatureToggle.enable!(:board_grant_effectuation_task)
```
9.  When reloading the page verify that the search query is in the url parameters and gets properly inserted into the search box
![image](https://user-images.githubusercontent.com/109369527/210840677-ad05fd7c-ce75-4c15-a980-c3236b2af5d2.png)

10. Backspace the name trem out of the search box and verify that the page state returns to the original page state without any search values entered
11.  Search for the value arch and it should return two records an HLR and SC
![image](https://user-images.githubusercontent.com/109369527/210841330-9ca57d82-fe97-4754-bc99-2b7e0e886bc5.png)

12. Search for the value john and verify that the pagination still works by clicking page 2 and 3
![image](https://user-images.githubusercontent.com/109369527/210842513-aeaf1adb-7852-4d68-8fa9-737d8d03b3f0.png)
![image](https://user-images.githubusercontent.com/109369527/210842573-14cbc4c1-2a4d-472b-9f2d-25a29d330dba.png)

13. Hit the x in the search bar and verify that the page returns to 1 and the default page load state with no search results
14. Search for the number 900 and verify that the participant ids that are returned include that number
![image](https://user-images.githubusercontent.com/109369527/210845074-0d74200a-9789-4053-a8d1-f9bca0dcca73.png)

15. Search for then number 84 and verify the results and that pagination still works
![image](https://user-images.githubusercontent.com/109369527/210845342-7a340432-dda9-4811-b10e-45733eeea3e1.png)
![image](https://user-images.githubusercontent.com/109369527/210845418-db776dcc-dfa0-4eca-bc0b-d36415e9c61c.png)

16.  In additional to all of the tests feel free to do manual intakes and search for those records as well to test for all of the various types of claimants and request issues.

17. The searching for the completed tab should behave in the exact same way as the in progress tab
```ruby
# if you are using my gzip data you might have to run the command below to get the data to show up if it has been 7 days passed when the data was created
BusinessLine.find_by(url:'vha').tasks.completed.each{ |t| t.update!(closed_at: 3.days.ago) }
```
18. On the completed tab search for fisher and verify the results
![image](https://user-images.githubusercontent.com/109369527/210867299-a0094dbd-c52e-4761-bbbd-5c0e6e6eda67.png)
19. On the completed tab search for the number 999 and verify the results
![image](https://user-images.githubusercontent.com/109369527/210867423-b3d26397-1a5d-45c0-90c5-a1d2e23f863b.png)
20. Use the x or the backspace and verify that the results are back to the original page state for the completed tab
![image](https://user-images.githubusercontent.com/109369527/210867578-001b4089-5220-48a5-a485-e8cd7767c571.png)
21. Verify any additional search or datatypes that you can think of that aren't mentioned in the testing plan.

#### Testing for 12466
1. Not sure yet maybe just load the page and see if there are still any database queries that are in addition to the main sql query now?
2. Do some render timing with and without the serializer changes to see how much it saves.




### Storybook Story
* [ ] NonCompTabs.stories.js

### Database Changes
*Only for Schema Changes*
- [ ] Nothing yet but might include some indexes
